### PR TITLE
[EC-507 / EC-508] SCIM CQRS Refactor - Groups/Users

### PR DESCRIPTION
**Pull request by r-tome**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Refactor the SCIM API to use the CQRS pattern which implies that each action must be isolated in a specific command to do only one specific task.
Added unit tests to check that each command completes its task.

## Code changes
All of these changes have been reviewed on other PRs:

* **bitwarden-server.sln:** Added new unit tests project Scim.Test
* **bitwarden_license/src/Scim/Groups/Interfaces/*:** SCIM group endpoints command/query interfaces
* **bitwarden_license/src/Scim/Groups/*:** SCIM group endpoints command/query implementations
* **bitwarden_license/src/Scim/Users/Interfaces/*:** SCIM user endpoints command/query interfaces
* **bitwarden_license/src/Scim/Users/*:** SCIM user endpoints command/query implementations
* **bitwarden_license/src/Scim/Controllers/v2/GroupsController.cs:** Updated to use commands and handle exceptions
* **bitwarden_license/src/Scim/Controllers/v2/UsersController.cs:** Updated to use commands and handle exceptions
* **bitwarden_license/src/Scim/Startup.cs:** Registered commands for injection
* **bitwarden_license/test/Scim.Test/Groups/*:** SCIM group command/query unit tests
* **bitwarden_license/test/Scim.Test/Users/*:** SCIM user command/query unit tests
* **bitwarden_license/test/Scim.Test/Scim.Test.csproj:** New unit test project for SCIM
* **bitwarden_license/test/Scim.Test/packages.lock.json:** Scim.Test project package dependencies
* **bitwarden_license/src/Scim/Utilities/ScimServiceCollectionExtensions.cs:** SCIM commands/queries dependency injection
* **src/Core/Exceptions/BadRequestException.cs:** Added parameterless constructor
* **src/Core/Exceptions/ConflictException.cs:** Added a new Exception type for Conflict
* **src/Core/Exceptions/NotFoundException.cs:** Added parameterless constructor
* **src/Core/OrganizationFeatures/OrganizationUsers/*:** OrganizationUser commands
* **src/Core/Services/Implementations/OrganizationService.cs:** Marked `DeleteUserAsync` as Obsolete
* **test/Core.Test/OrganizationFeatures/OrganizationUsers/DeleteOrganizationUserCommandTests.cs:** Command unit tests

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Core/OrganizationFeatures/OrganizationUsers/DeleteGroupCommand.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿using Bit.Core.Exceptions;
```
Comment (eliykat) [inline src/Core/OrganizationFeatures/OrganizationUsers/DeleteGroupCommand.cs:1]: I only just noticed this is under `OrganizationUsers`, but Groups are a different feature.

I suggest creating a new `OrganizationFeatures/Groups` folder.

**Review by eliykat — CHANGES_REQUESTED** [review]
One minor thing I noticed, otherwise great work on this and thanks for your patience working through it all!

File: src/Core/OrganizationFeatures/OrganizationUsers/DeleteGroupCommand.cs (hunk 1 above)
Comment (r-tome) [reply src/Core/OrganizationFeatures/OrganizationUsers/DeleteGroupCommand.cs]: Makes sense! I've made that change.

**Review by eliykat — APPROVED** [review]

**Comment by shane-melton** [comment]
@r-tome @eliykat 

For the Group commands, do we need to label them as "SCIM specific" in some way? Something like `IPostScimGroupCommand`?

 I am in the process of converting a few non-SCIM group controller endpoints over to the CQRS pattern [here](https://github.com/bitwarden/server/pull/2280/files#diff-1ce7af6c9a3502f7b0eae80c2c7220f502ee7e4a4776fb9862f23907647b700c) and have some new commands with similar names.

I see that they depend on ScimRequest models and have some SCIM only logic in them. I don't believe we'd be able to use these commands in the [non-SCIM GroupController](https://github.com/bitwarden/server/blob/8f37a5119cf8199aa193bb1fcd461d7b5c4ab666/src/Api/Controllers/GroupsController.cs) as it would require casting to the Scim request models and not all organizations would have a ScimContext.

https://github.com/bitwarden/server/blob/8f37a5119cf8199aa193bb1fcd461d7b5c4ab666/bitwarden_license/src/Scim/Groups/Interfaces/IPostGroupCommand.cs#L4-L9
https://github.com/bitwarden/server/blob/8f37a5119cf8199aa193bb1fcd461d7b5c4ab666/bitwarden_license/src/Scim/Groups/PostGroupCommand.cs#L47-L52

**Comment by eliykat** [comment]
> I don't believe we'd be able to use these commands in the [non-SCIM GroupController](https://github.com/bitwarden/server/blob/8f37a5119cf8199aa193bb1fcd461d7b5c4ab666/src/Api/Controllers/GroupsController.cs)

Yes that's right, everything in the `Scim` project is Scim-specific. Any generally useful commands were added to `Core` instead (`DeleteGroupCommand` and `DeleteOrganizationUserCommand`), but there weren't many of those.

I thought that them being in a separate project and namespace would be enough to distinguish them from commands in `Core`, but I honestly don't know if there's a strong naming convention either way.

**Comment by shane-melton** [comment]
> I thought that putting the Scim-specific classes in a separate project and namespace would be enough to distinguish them from commands and queries in `Core`, but I honestly don't know if there's a strong naming convention either way.

Ah, you're totally right, I agree that the separate namespace should be enough. I just missed that when I was focusing on the `DeleteGroupCommand` in `Core`.

@r-tome Please ignore my comment and nice work on the refactor!

**Pull request merged**